Catch errors during android plugin registration

atrope opened the pull request:
This PR Apply the changes explained in #75416

It does not change anything but the order on which we are registering the plugins and (at least temporarily) fix issues that are occurring when using V2 Plugins that are alphabetically being registered after any other V1 Plugin on Isolates.

This can potentially avoid many users having bad experiences and losing many hours trying to find the problem.


- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing. 

_admin edit:_
fixes https://github.com/flutter/engine/pull/25496

atrope committed:
sorting android plugins V2->V1

atrope committed:
fixing strong analyzer syntax

jmagman commented:
This needs tests, take a look at https://github.com/flutter/flutter/blob/515ef91ba1f6021403e30ecf535a852e3cbd9a2c/packages/flutter_tools/test/general.shard/plugins_test.dart#L699.  You could validate that the contents of GeneratedPluginRegistrant.java are sorted.

atrope committed:
adding tests

atrope commented:
Hi @jmagman Thanks for the tip! Added the test :).

jonahwilliams reviewed (commented):
One caveat is that we're already sorting the plugins much earlier in the process, but putting them strictly in alphabetical order. I'm not sure if sorting when we're writing out the android plugin registrant is the correct choice, since we only have untyped data - what if the structure changes slightly but our unit tests don't hit it?

Did you investigate sorting the plugin list earlier?

atrope commented:
Hey @jonahwilliams we could potentially do that before as it is doing the alphabetical order:

```
  plugins.sort((Plugin left, Plugin right) => left.name.compareTo(right.name));
  if (androidPlatform) {
    await _writeAndroidPluginRegistrant(project, plugins);
  }

```

But we are using now the platform map that is done in the _extractPlatformMaps function  the line before inside _writeAndroidPluginRegistrant where we are doing the sort.

Let me know if there is a parameter before that i can use and i will gladly change to sort before.

jonahwilliams commented:
I don't know that I like this solution TBH, double sorting is going to be confusing. For my read of the issue linked, it sounds like there was a problem with a particular plugin, and that was preventing anything after that plugin from being registered.

If that is the case, would a more reasonable fix be to add a try/catch around each plugin registration? FYI @xster

atrope commented:
As I wrote in the issue this is indeed  a workaround to be able to use V2 and V1 plugins in the app without breaking the automatic registration for v2 plugins and be able to use them in a background service.

jonahwilliams commented:
ack. In that case your change makes sense but I think there is a better location for it, while we still have the typed Plugin objects and not maps. let me take another look real quick

jonahwilliams commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -620,7 +620,12 @@ AndroidEmbeddingVersion _getAndroidEmbeddingVersion(FlutterProject project) {
 Future<void> _writeAndroidPluginRegistrant(FlutterProject project, List<Plugin> plugins) async {
   final List<Map<String, dynamic>> androidPlugins =
```
How about instead of sorting the untyped map, create a new List of plugins here and apply the sort on the typed Plugin objects:

```
final List<Plugin> ordered = plugins.toList()
  ..sort(...);
```

atrope replied:
I don't know if I undestane but I will take a look and update here when I can!

atrope replied:
@jonahwilliams i checked it out and this does not make sense because _extractPlatformMaps also removes all plugins not related to Android and i can only "remove" the others inside the _writeAndroidPluginRegistrant function.  
For example:

```
url_launcher - {name: url_launcher, package: io.flutter.plugins.urllauncher, class: UrlLauncherPlugin, supportsEmbeddingV1: true, supportsEmbeddingV2: true}
url_launcher_linux - null
url_launcher_macos - null
url_launcher_web - null
url_launcher_windows - null
video_player - {name: video_player, package: io.flutter.plugins.videoplayer, class: VideoPlayerPlugin, supportsEmbeddingV1: true, supportsEmbeddingV2: true}
video_player_web - null
```

Also, I Only know if the plugin is V2 or not inside the "toMap" function.

  ```
@override
  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      'name': name,
      'package': package,
      'class': pluginClass,
      // Mustache doesn't support complex types.
      'supportsEmbeddingV1': _supportedEmbedings.contains('1'),
      'supportsEmbeddingV2': _supportedEmbedings.contains('2'),
    };
  }
```

What I could do  is to expose a new method in the AndroidPlugin class that would tell me if the plugin is V2 or not.

But then i would need to check  if the plugin is Android(which is done inside the `_extractPlatformMaps`) and ifit is V2.

IMHO we should use the Untyped map instead of double checking twice the platform.

LMK what you think!

blasten commented:
@xster is this due to the R8 issue that you already fixed?

xster commented:
Good call @blasten, thanks for pointing out. It could be. We recently had this issue where some plugins' dependent classes can't classload while registering. This causes that plugin and all subsequent plugins to not register. You might be observing a side effect of it where plugins registered earlier will work and later plugins (especially after a crashing one) don't. Besides that, there shouldn't be any intrinsic differences or ordering requirements between v1 and v2 plugins. Generally, we should find and resolve issues where plugins creation calls fail during registration. 

Tangentially, we can also increase the resilience of the system when a plugin does fail. But rather than sorting them, it might be better, in the same part of code, where the GeneratedPluginRegistrant is generated, to wrap each plugin registration call with a try catch, so plugins subsequent to a crashing plugin can still work.

atrope commented:
Hi @xster and @blasten, 

1 - I Just tested with the master branch and the issue is still  happening, it only changed the error message, but still can't find the plugin because a V1 plugin broke the registration..

```
I/flutter (31312): FlutterFire Messaging: An error occurred in your background messaging handler:
I/flutter (31312): MissingPluginException(No implementation found for method getDatabasesPath on channel com.tekartik.sqflite)

```


```
Flutter 2.1.0-13.0.pre.416 • channel master • https://github.com/flutter/flutter.git
Framework • revision b9638fb837 (8 hours ago) • 2021-04-02 00:39:02 -0400
Engine • revision bed3a4a7a0
Tools • Dart 2.13.0 (build 2.13.0-192.0.dev)
```

2 - Do You think it is better to add a try catch for each plugin instead of sorting them? I Already tried and it also worked. Which one do you think is the best strategy?

atrope commented:
I Tried something like this:
```

const String _androidPluginRegistryTemplateNewEmbedding = '''
package io.flutter.plugins;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;
import android.util.Log;

import io.flutter.embedding.engine.FlutterEngine;
{{#needsShim}}
import io.flutter.embedding.engine.plugins.shim.ShimPluginRegistry;
{{/needsShim}}

/**
 * Generated file. Do not edit.
 * This file is generated by the Flutter tool based on the
 * plugins that support the Android platform.
 */
@Keep
public final class GeneratedPluginRegistrant {
  private static final String TAG = "GeneratedPluginRegistrant";
  public static void registerWith(@NonNull FlutterEngine flutterEngine) {
{{#needsShim}}
    ShimPluginRegistry shimPluginRegistry = new ShimPluginRegistry(flutterEngine);
{{/needsShim}}
{{#plugins}}
  {{#supportsEmbeddingV2}}
    try{
      flutterEngine.getPlugins().add(new {{package}}.{{class}}());
    } catch(Exception e) { 
      Log.i(TAG, "Error registering {{package}}.{{class}}"); 
    }
  {{/supportsEmbeddingV2}}
  {{^supportsEmbeddingV2}}
    {{#supportsEmbeddingV1}}
     try{ 
       {{package}}.{{class}}.registerWith(shimPluginRegistry.registrarFor("{{package}}.{{class}}"));
      } catch(Exception e) {
        Log.i(TAG, "Error registering {{package}}.{{class}}");
      }
    {{/supportsEmbeddingV1}}
  {{/supportsEmbeddingV2}}
{{/plugins}}
  }
}
''';
```

blasten commented:
Wrapper try/catch. As a separate note, we should find out why those plugins are broken.

blasten commented:
In https://github.com/flutter/flutter/pull/78964#issuecomment-812522788, I would change `Log.i` for `Log.e` and I would also log the exception that you caught.  Something along these lines:

```java
 Log.e(TAG, "Error registering plugin `{{name}}`, package `{{package}}.{{class}}`. " +
            "The app might not function as expected unless you remove it from pubspec.yaml. " +
            "Exception:" + Log.getStackTraceString(e));
```

xster commented:
nit: it might print a bit nicer using the Log.e(tag, message, exception) signature since there's already built-in support for taking an exception

xster commented:
If you do @blasten's suggestion, you'll end up fixing https://github.com/flutter/flutter/issues/78258 :)

atrope commented:
Will update this PR with all the suggested changes.

atrope committed:
changing sort to try catch

atrope committed:
Logging Exception

atrope committed:
restoring deleted test

atrope commented:
Hey @xster, @blasten  and @jonahwilliams 

Updated the template to include the try catch and removed the sort as it is not needed.

Now when the V1 plugin is registering it is showing the error but continue with the app as it should :)

```
I/FLTFireBGExecutor(26306): Creating background FlutterEngine instance, with args: [--start-paused, --enable-dart-profiling]
E/GeneratedPluginRegistrant(26306): Error registering plugin mdns_plugin, com.mutablelogic.mdns_plugin.MDNSPlugin The app might not function as expected unless you remove it from pubspec.yaml.
E/GeneratedPluginRegistrant(26306): java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.Object android.app.Activity.getSystemService(java.lang.String)' on a null object reference
E/GeneratedPluginRegistrant(26306): 	at com.mutablelogic.mdns_plugin.MDNSPlugin.<init>(MDNSPlugin.kt:52)
E/GeneratedPluginRegistrant(26306): 	at com.mutablelogic.mdns_plugin.MDNSPlugin$Companion.registerWith(MDNSPlugin.kt:29)
E/GeneratedPluginRegistrant(26306): 	at com.mutablelogic.mdns_plugin.MDNSPlugin.registerWith(Unknown Source:2)
E/GeneratedPluginRegistrant(26306): 	at io.flutter.plugins.GeneratedPluginRegistrant.registerWith(GeneratedPluginRegistrant.java:165)
E/GeneratedPluginRegistrant(26306): 	at java.lang.reflect.Method.invoke(Native Method)
E/GeneratedPluginRegistrant(26306): 	at io.flutter.embedding.engine.plugins.util.GeneratedPluginRegister.registerGeneratedPlugins(GeneratedPluginRegister.java:34)
E/GeneratedPluginRegistrant(26306): 	at io.flutter.embedding.engine.FlutterEngine.<init>(FlutterEngine.java:345)
E/GeneratedPluginRegistrant(26306): 	at io.flutter.embedding.engine.FlutterEngine.<init>(FlutterEngine.java:260)
E/GeneratedPluginRegistrant(26306): 	at io.flutter.embedding.engine.FlutterEngine.<init>(FlutterEngine.java:240)
E/GeneratedPluginRegistrant(26306): 	at io.flutter.embedding.engine.FlutterEngine.<init>(FlutterEngine.java:160)
E/GeneratedPluginRegistrant(26306): 	at io.flutter.plugins.firebase.messaging.FlutterFirebaseMessagingBackgroundExecutor.lambda$null$0$FlutterFirebaseMessagingBackgroundExecutor(FlutterFirebaseMessagingBackgroundExecutor.java:185)
E/GeneratedPluginRegistrant(26306): 	at io.flutter.plugins.firebase.messaging.-$$Lambda$FlutterFirebaseMessagingBackgroundExecutor$Tw1_-iY8wcRJPWf1qeylRYPPEZI.run(Unknown Source:6)
E/GeneratedPluginRegistrant(26306): 	at android.os.Handler.handleCallback(Handler.java:938)
E/GeneratedPluginRegistrant(26306): 	at android.os.Handler.dispatchMessage(Handler.java:99)
E/GeneratedPluginRegistrant(26306): 	at android.os.Looper.loop(Looper.java:246)
E/GeneratedPluginRegistrant(26306): 	at android.app.ActivityThread.main(ActivityThread.java:8456)
E/GeneratedPluginRegistrant(26306): 	at java.lang.reflect.Method.invoke(Native Method)
E/GeneratedPluginRegistrant(26306): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:596)
E/GeneratedPluginRegistrant(26306): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
.
.
.
D/SocialShare(26306): onAttachedToEngine
```


First I Thought on creating a function and pass the name, class and package to the function and Log from it do you think its better to concentrate everything in a function or directly in the catch as it is now?

atrope committed:
Merge branch 'master' into sort-v2-v1

atrope committed:
update from master

atrope commented:
Hey Guys,
Does anyone else have any input?
If not can this be merged so we can use without the need to modify our local flutter copy? ;)

blasten reviewed (approved):
LGTM

xster commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -567,6 +567,7 @@ package io.flutter.plugins;
 
 import androidx.annotation.Keep;
 import androidx.annotation.NonNull;
+import android.util.Log;
```
please import io.flutter.Log here

xster commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -580,17 +581,28 @@ import io.flutter.embedding.engine.plugins.shim.ShimPluginRegistry;
+      flutterEngine.getPlugins().add(new {{package}}.{{class}}());
+    } catch(Exception e) {
+      Log.e(TAG, "Error registering plugin {{name}}, {{package}}.{{class}} " +
+      "The app might not function as expected unless you remove it from pubspec.yaml.", e);
```
I'd avoid giving an official recommendation here since it may be "fixable" depending on the usage of the plugin itself. Just keeping the first sentence is fine.

xster reviewed (commented):
Great. This is much better! 

Please also create a test in https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/general.shard/plugins_test.dart to verify the content of the new GeneratedPluginRegistrant.java. The test can be modeled after an existing test.

atrope committed:
Applying review notes and Adding test

atrope commented:
@xster applied the review notes and added the tests to ensure the plugins are wrapped in try catch, please take a look

xster reviewed (approved):
Excellent. Thank you for working through the comments and for your contribution!
